File: app/src/main/java/com/nativo/sampleapp/ViewAdapter/GridViewAdapter.java
package com.nativo.sampleapp.ViewAdapter;

import static com.nativo.sampleapp.util.AppConstants.CLICK_OUT_URL;
import static com.nativo.sampleapp.util.AppConstants.SECTION_URL;
import static com.nativo.sampleapp.util.AppConstants.SP_CAMPAIGN_ID;
import static com.nativo.sampleapp.util.AppConstants.SP_CONTAINER;
import static com.nativo.sampleapp.util.AppConstants.SP_SECTION_URL;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.net.Uri;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.GridView;
import android.widget.ImageView;
import android.widget.TextView;
import com.nativo.sampleapp.R;
import net.nativo.sdk.NativoSDK;
import net.nativo.sdk.NtvAdData;
import net.nativo.sdk.NtvSectionAdapter;
import net.nativo.sdk.injectable.NtvInjectable;
import net.nativo.sdk.injectable.NtvInjectableType;

import java.util.ArrayList;
import java.util.List;

public class GridViewAdapter extends BaseAdapter {

    private GridView gridView;
    List<Integer> integerList = new ArrayList<>();

    public GridViewAdapter(Context context, GridView gridView) {
        this.gridView = gridView;
        for (int i = 0; i < 10; i++) {
            integerList.add(i);
        }

    }

    @Override
    public int getCount() {
        return integerList.size();
    }

    @Override
    public Object getItem(int i) {
        return null;
    }

    @Override
    public long getItemId(int i) {
        return 0;
    }

    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        // Publisher article view
        if (view == null) {
            view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.publisher_article, viewGroup, false);
        }
        bindView(view, i);
        return view;
    }

    private void bindView(View view, int i) {
        if (view != null) {
            if (((ImageView) view.findViewById(R.id.article_image)) != null) {
                ((ImageView) view.findViewById(R.id.article_image)).setImageResource(R.drawable.newsimage);
            }
            if (((ImageView) view.findViewById(R.id.sponsored_ad_indicator)) != null) {
                ((ImageView) view.findViewById(R.id.sponsored_ad_indicator)).setVisibility(View.INVISIBLE);
            }
            if (((TextView) view.findViewById(R.id.article_author)) != null) {
                ((TextView) view.findViewById(R.id.article_author)).setText(R.string.sample_author);
            }
            if (((TextView) view.findViewById(R.id.article_title)) != null) {
                ((TextView) view.findViewById(R.id.article_title)).setText(R.string.sample_title);
            }
            view.setOnClickListener(onClickListener);
        }
    }

    View.OnClickListener onClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            view.getContext().startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(CLICK_OUT_URL)));
        }
    };

}

File: app/src/main/java/com/nativo/sampleapp/ViewAdapter/GridViewAdapter.kt
package com.nativo.sampleapp.ViewAdapter

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.nativo.sampleapp.R
import com.nativo.sampleapp.util.AppConstants
import com.nativo.sampleapp.util.AppConstants.NtvTAG
import net.nativo.sdk.*
import net.nativo.sdk.NtvAdData
import net.nativo.sdk.injectable.NtvInjectable
import net.nativo.sdk.injectable.NtvInjectableType

/**
 * Example of Nativo SDK implemented using [GridView]
 * Ads are placed according to rule in link`shouldPlaceAdAtIndex()`.
 * If an ad is not placed(eg no fill scenario) the cell is marked with red
 */
class GridViewAdapter(
    val context: Context, private val gridView: GridView
) : BaseAdapter(), NtvSectionAdapter {

    companion object {
        const val ITEM_COUNT = 20
    }

    private val sudoArticleList = mutableListOf<String>()
    private val nativoAdIndexes = mutableListOf(1, 3, 5, 7, 9, 11, 13, 15)
    private var initialNativoRequest = true

    // Create unique layout class for Nativo ads
    private class NativoLayout(context: Context) : FrameLayout(context)

    /**
     * Show [AppConstants.CLICK_OUT_URL]
     */
    private var itemClickListener = Runnable {
        context.startActivity(
            Intent(
                Intent.ACTION_VIEW,
                Uri.parse(AppConstants.CLICK_OUT_URL)
            )
        )
    }

    init {
        // Nativo init
        NativoSDK.initSectionWithAdapter(this, AppConstants.SECTION_URL, context)

        for (i in 0..(ITEM_COUNT + nativoAdIndexes.size)) {
            sudoArticleList.add("Article $i")
        }
    }

    override fun getCount(): Int {
        return sudoArticleList.size
    }

    override fun getItem(p0: Int): Any? {
        return null
    }

    override fun getItemId(p0: Int): Long {
        return 0
    }

    override fun getView(index: Int, view: View?, viewGroup: ViewGroup): View {
        val cellView: View
        if (nativoAdIndexes.contains(index)) {
            // We need a blank view for Nativo ads.
            cellView = if (view !is NativoLayout) NativoLayout(context) else view
            NativoSDK.placeAdInView(cellView, gridView, AppConstants.SECTION_URL, index, null)
        } else if (view == null || view is NativoLayout) {
            // Create new publisher article view
            cellView = LayoutInflater.from(context).inflate(R.layout.publisher_article, viewGroup, false)
            bindView(cellView, sudoArticleList[index])
        } else {
            cellView = view
            bindView(cellView, sudoArticleList[index])
        }
        return cellView
    }

    @SuppressLint("SetTextI18n")
    private fun bindView(view: View, item: String) {
        val articleImage: ImageView? = view.findViewById(R.id.article_image)
        val articleTitle: TextView? = view.findViewById(R.id.article_title)
        val articleAuthor: TextView? = view.findViewById(R.id.article_description)

        articleImage?.setImageResource(R.drawable.newsimage)
        articleAuthor?.setText(R.string.sample_author)
        articleTitle?.text = item

        view.setOnClickListener {
            itemClickListener.run()
        }
    }

    override fun didReceiveAd(didGetFill: Boolean, inSection: String) {
        Log.d(NtvTAG, "Did receive ad: $didGetFill")

        if (didGetFill && initialNativoRequest) {
            Log.w(NtvTAG, "Needs Reload Everything")
            notifyDataSetChanged()
        }

        initialNativoRequest = false
    }

    override fun didAssignAdToLocation(
        location: Int,
        adData: NtvAdData,
        inSection: String,
        container: ViewGroup
    ) {
        Log.d(NtvTAG, "didAssignAdToLocation: $location")
        notifyDataSetChanged()
    }

    override fun didPlaceAdInView(
        view: View,
        adData: NtvAdData,
        injectable: NtvInjectable,
        atLocation: Int,
        inSection: String,
        container: ViewGroup
    ) {
        Log.d(NtvTAG, "didPlaceAdInView: $atLocation AdData: $adData")
    }

    override fun didFailAd(
        inSection: String,
        atLocation: Int?,
        inView: View?,
        container: ViewGroup?,
        error: Throwable?
    ) {
        Log.d(NtvTAG, "onFail at location: $atLocation Error: $error")
        if (atLocation != null) {
            Log.w(NtvTAG, "Removing Nativo Ad!")
            sudoArticleList.removeAt(atLocation)
            nativoAdIndexes.remove(atLocation);
            notifyDataSetChanged()
        }
    }

    override fun needsDisplayLandingPage(
        landingPageIntent: Intent,
        inSection: String,
        container: ViewGroup
    ) {
        context.startActivity(landingPageIntent)
    }

    override fun needsDisplayClickOutURL(url: String, inSection: String, container: ViewGroup) {
        context.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
    }

    override fun <T : NtvInjectable> registerInjectableClassForTemplateType(
        injectableType: NtvInjectableType,
        atLocation: Int?,
        inSection: String?
    ): Class<T>? {
        return null
    }
}